### DISCUSSION TITLE
RSA keys changed from der to pem, now new vaultwarden can't read old RSA keys anymore

### DISCUSSION THREAD

**Discussion by eriklogtenberg** [dani-garcia/vaultwarden#2193] [Q&A] [answered]
### Subject of the issue
After upgrade from Vaultwarden 1.21.0 to 1.23.1, server is unable to start. Error message is:
"[vaultwarden][ERROR] Error creating keys, exiting..."

Tracked it down to this commit: 46e0f3c43a81ce9411612c152e414162a9c220ac, where the default RSA key format (and filename) is changed from DER to PEM. It seems that as a result, the existing DER files can no longer be found.
What is the correct upgrade path?

### Deployment environment

* Install method: official Docker image from https://hub.docker.com/r/vaultwarden/server

**Comment by BlackDex** [comment, marked as answer]
It shouldn't be an issue, at all. But you may just remove all those files and new ones will be generated.
It will only only hurt current sessions and invite links which use those files.

**Comment by louPxL** [comment]
Same issue here.

docker run -v $PWD/data-bitwarden:/data -p 80:80 vaultwarden/server:1.24.0

[INFO] No .env file found.
[2022-03-10 20:57:19.399][vaultwarden][ERROR] Error creating keys, exiting...


In my Desktop Fedora 35,
But works like a charm on my Debian Bullseye Laptop

FYI : same result if folder data-bitwarden is new or not

**Reply by ghost** [reply]
I know it's a late reply and this is probably already solved, but in case anyone else is looking, It was selinux causing the issue for me. 

Confirmed with "`seteneforce 0`" then running the container. 
Once confirmed, turn selinux back on with "`setenforce 1`".
Now label your file system accordingly:
```
semanage fcontext -a -t container_file_t "/my-vaultwarden-data/(/.*)?"
restorecon -R $PWD/my-vaultwarden-data
docker run -v /my-vaultwarden-data:/data -p 80:80 vaultwarden/server:1.24.0
```

**Reply by Chinoman10** [reply]
I don't have SELinux enforced and am still running into this problem.
Even after removing all the certificate files, and giving full permissions to the folder...

I'm out of ideas.

**Comment by iamfelix7** [comment]
Same issue here..

I'm using the latest docker image.

[INFO] No .env file found.
[2023-01-19 19:08:23.298][vaultwarden::util][ERROR] Can't create 'data/rsa_key.pem': Permission denied
[2023-01-19 19:08:23.298][vaultwarden][ERROR] Error creating keys, exiting...

Should I remove the "rsa_key.pem" file? Where can I find this file?

Or is there a problem with the permissions?
drwxr-xr-x  15 nobody nogroup  4096 Jan 19 00:00 vw-data

Thank you guys!

**Reply by stefan0xC** [reply]
> [2023-01-19 19:08:23.298][vaultwarden::util][ERROR] Can't create 'data/rsa_key.pem': **Permission denied** 
> *[...]*
> Or is there a problem with the permissions? drwxr-xr-x 15 nobody nogroup 4096 Jan 19 00:00 vw-dvw-data
>

Looks like nobody has **w**rite access to the directory so vaultwarden could not create the file. Is the `vw-data` folder empty or did you update from an older version?

**Reply by iamfelix7** [reply]
Hi Stefan!

It seems to be a permission problem. The `vw-data` folder is not empty - I installed an update 2 months ago. But the error persists since a few days. I also tried to change the owner of the vw-data folder with chown command without success..

Maybe someone has an idea how i can get the permissions adjusted again?

**Reply by BlackDex** [reply]
How do you start Vaultwarden? Are you using docker? If so, is it configured to run as non-root? Could you show us the docker command or docker-compose file you use to run it (With all the secrets removed of course).

Within the container, Vaultwarden is run as `root`, and thus those files need to be owned by root, and not nobody/nogroup.

**Reply by iamfelix7** [reply]
Yes, I'm using docker.

I just use the default command `docker run -d --name vaultwarden -v /vw-data/:/data/ -p 80:80 vaultwarden/server:latest`

I have to check if it is configured to run as non-root.

**Reply by BlackDex** [reply]
Well, you are using port 80, so non-root is out of the question.
I suggest to chown those files and folder to be owned by root.
`chown -R root:root /vw-data`

**Reply by iamfelix7** [reply]
I restored a backup from november 2022 (vaultwarden was running flawless at this time!) and strangely enough, exactly the same problem exists here..

**Reply by iamfelix7** [reply]
@BlackDex thank you for the tip! I already tried to adjust the permission with `chown -R root:root /vw-data` - the operation is not permitted - I run the command as root-user, so I don't know why the operation is not permitted :(

**Reply by BlackDex** [reply]
Are you running this inside an other container? Shared storage or something?

**Reply by iamfelix7** [reply]
Setup:
 - Proxmox VE 7.2-3 Host
 - LXC Container running on Proxmox Host
 - inside this LXC Container runs the Vaultwarden container (since 8 months without any problem, until now)

**Reply by BlackDex** [reply]
Then probably you have the `/vw-data` mounted from the host, and lxc controls those rights.

Probably a restart or reboot changed some settings. But it has nothing to do with Vaultwarden updates that those rights are changed.

So, i suggest to check the Proxmox settings.

**Reply by iamfelix7** [reply]
Thanks for your support BlackDex! I also think it is not a vaultwarden problem! I will update my Proxmox host now and try it again....
